File: TrafficControl/src/trafficcontrol/Controller.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package trafficcontrol;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 *
 * @author owner
 */
public class Controller {
    
    static Lane [][] ArrayOfLanes = new Lane[4][2];  // all lanes
    static Lane L1;
    static Lane L2;
    static int Qmax=0;
    
    // Valid cases initilization
    
    
    
        public int getQmax(){
            for(int i=0; i<4; i++){
                for(int j=0; j<2; j++){
                    if(ArrayOfLanes[i][j].Q > Qmax){
                        Qmax = ArrayOfLanes[i][j].Q;
                        L1 = ArrayOfLanes[i][j];
                    }
            }
        }
        return Qmax;
    }
            
    public static Lane getL2(Lane L1){
        
        int QlocalMax=0;
        if(L1.ascL1.Q > QlocalMax){
        QlocalMax = L1.ascL1.Q;
        L2 = L1.ascL1;
        }
        if(L1.ascL2.Q > QlocalMax){
        QlocalMax = L1.ascL2.Q;
        L2 = L1.ascL2;
        }
        if(L1.ascL3.Q > QlocalMax){
        QlocalMax = L1.ascL3.Q;
        L2 = L1.ascL3;
        }
        
        return L2;
}
    public static void lanesInit(){
        for(int i=0; i<4; i++){
            for(int j=0; j<2; j++){
                Controller.ArrayOfLanes[i][j]= new Lane(i,j);
            }
        }
    }

    public static void casesInit (){
        // N S E w  F R
        // 0 1 2 3  0 1      
        
        // N,F
       ArrayOfLanes[0][0].ascL1 = ArrayOfLanes[0][1];
       ArrayOfLanes[0][0].ascL2 = ArrayOfLanes[1][0];
       ArrayOfLanes[0][0].ascL3 = ArrayOfLanes[3][1];
        
        // S,F
       ArrayOfLanes[1][0].ascL1 = ArrayOfLanes[1][1];
       ArrayOfLanes[1][0].ascL2 = ArrayOfLanes[2][1];
       ArrayOfLanes[1][0].ascL3 = ArrayOfLanes[0][0];
        
        // E,F
       ArrayOfLanes[2][0].ascL1 = ArrayOfLanes[2][1];
       ArrayOfLanes[2][0].ascL2 = ArrayOfLanes[3][0];
       ArrayOfLanes[2][0].ascL3 = ArrayOfLanes[0][1];
        
        // W,F
       ArrayOfLanes[3][0].ascL1 = ArrayOfLanes[3][1];
       ArrayOfLanes[3][0].ascL2 = ArrayOfLanes[1][1];
       ArrayOfLanes[3][0].ascL3 = ArrayOfLanes[2][0];
        
        // N,R
       ArrayOfLanes[0][1].ascL1 = ArrayOfLanes[0][0];
       ArrayOfLanes[0][1].ascL2 = ArrayOfLanes[1][1];
       ArrayOfLanes[0][1].ascL3 = ArrayOfLanes[2][0];
        
        // S,R
       ArrayOfLanes[1][1].ascL1 = ArrayOfLanes[1][0];
       ArrayOfLanes[1][1].ascL2 = ArrayOfLanes[0][1];
       ArrayOfLanes[1][1].ascL3 = ArrayOfLanes[3][0];
        
        // E,R
       ArrayOfLanes[2][1].ascL1 = ArrayOfLanes[2][0];
       ArrayOfLanes[2][1].ascL2 = ArrayOfLanes[3][1];
       ArrayOfLanes[2][1].ascL3 = ArrayOfLanes[1][0];
        
        // W,R
       ArrayOfLanes[3][1].ascL1 = ArrayOfLanes[3][0];
       ArrayOfLanes[3][1].ascL2 = ArrayOfLanes[2][1];
       ArrayOfLanes[3][1].ascL3 = ArrayOfLanes[0][0];
                
    }
}

File: Traffic control/Controller.py
import Lane


class Controller(object):
    arrayoflanes = [[Lane.Lane(0, 0) for j in range(2)] for i in range(4)]
    L1 = Lane.Lane(0, 0)
    L2 = Lane.Lane(0, 0)
    Qmax = 0

    @staticmethod
    def getqmax():
        for i in range(4):
            for j in range(2):
                if Controller.arrayoflanes[i][j].Q > Controller.Qmax:
                    Controller.Qmax = Controller.arrayoflanes[i][j].Q
                    Controller.L1 = Controller.arrayoflanes[i][j]

        return Controller.Qmax

    @staticmethod
    def getl2(l1):

        qlocalmax = 0
        if l1.ascl1.Q > qlocalmax:
            qlocalmax = l1.ascl1.Q
        Controller.L2 = l1.ascl1

        if l1.ascl2.Q > qlocalmax:
            qlocalmax = l1.ascl2.Q
        Controller.L2 = l1.ascl2

        if l1.ascl3.Q > qlocalmax:
            qlocalmax = l1.ascl3.Q
        Controller.L2 = l1.ascl3

        return Controller.L2

    @staticmethod
    def lanesinit():
        for i in range(4):
            for j in range(2):
                Controller.arrayoflanes[i][j] = Lane.Lane(i, j)
                Controller.arrayoflanes[i][j].D = i
                Controller.arrayoflanes[i][j].L = j

    @staticmethod
    def casesinit():
        # N S E w  F R
        # 0 1 2 3  0 1

        # N,F
        Controller.arrayoflanes[0][0].ascl1 = Controller.arrayoflanes[0][1]
        Controller.arrayoflanes[0][0].ascl2 = Controller.arrayoflanes[1][0]
        Controller.arrayoflanes[0][0].ascl3 = Controller.arrayoflanes[3][1]

        # S,F
        Controller.arrayoflanes[1][0].ascl1 = Controller.arrayoflanes[1][1]
        Controller.arrayoflanes[1][0].ascl2 = Controller.arrayoflanes[2][1]
        Controller.arrayoflanes[1][0].ascl3 = Controller.arrayoflanes[0][0]

        # E,F
        Controller.arrayoflanes[2][0].ascl1 = Controller.arrayoflanes[2][1]
        Controller.arrayoflanes[2][0].ascl2 = Controller.arrayoflanes[3][0]
        Controller.arrayoflanes[2][0].ascl3 = Controller.arrayoflanes[0][1]

        # W,F
        Controller.arrayoflanes[3][0].ascl1 = Controller.arrayoflanes[3][1]
        Controller.arrayoflanes[3][0].ascl2 = Controller.arrayoflanes[1][1]
        Controller.arrayoflanes[3][0].ascl3 = Controller.arrayoflanes[2][0]

        # N,R
        Controller.arrayoflanes[0][1].ascl1 = Controller.arrayoflanes[0][0]
        Controller.arrayoflanes[0][1].ascl2 = Controller.arrayoflanes[1][1]
        Controller.arrayoflanes[0][1].ascl3 = Controller.arrayoflanes[2][0]

        # S,R
        Controller.arrayoflanes[1][1].ascl1 = Controller.arrayoflanes[1][0]
        Controller.arrayoflanes[1][1].ascl2 = Controller.arrayoflanes[0][1]
        Controller.arrayoflanes[1][1].ascl3 = Controller.arrayoflanes[3][0]

        # E,R
        Controller.arrayoflanes[2][1].ascl1 = Controller.arrayoflanes[2][0]
        Controller.arrayoflanes[2][1].ascl2 = Controller.arrayoflanes[3][1]
        Controller.arrayoflanes[2][1].ascl3 = Controller.arrayoflanes[1][0]

        # W,R
        Controller.arrayoflanes[3][1].ascl1 = Controller.arrayoflanes[3][0]
        Controller.arrayoflanes[3][1].ascl2 = Controller.arrayoflanes[2][1]
        Controller.arrayoflanes[3][1].ascl3 = Controller.arrayoflanes[0][0]
